Write unique game info to /tmp/GAMENAME

misteraddons opened the pull request:
This is the game equivalent of writing the current core to /tmp/CORENAME. 

The idea was inspired by Robert Peip's PSX and N64 cores, which identify internal titles for games.

This new function writes the path, game_id (if applicable) and crc32 (if applicable) to /tmp/GAMENAME.
`user_io_write_gamename(const char *path, const char *game_id, uint32_t crc32_val);`

- Cartridge games are identified already identified by crc32
  - Special cases
    - N64 - crc32 and internal title (already done by Robert Peip)
    - NeoGeo - internal title instead of crc32
- Disc games are identified in unique ways (internal title)
  - Tested
    - MegaCD - internal title
    - NeoGeo CD - disc author date + internal title
    - CD-I - internal title
    - PSX - internal title (already done by Robert Peip)
    - Saturn - internal title
    - TGFX16-CD - md5 hash of first 100 sectors
- MGL files work as expected
- MRA files export the mame setname as game_id
- Computer support may vary

Inspired by @niemasd's [GameID](https://github.com/niemasd/GameID) project.

misteraddons committed:
Write unique game info to /tmp/GAMENAME

sorgelig commented:
Looking at the amount of code added with a lot of heuristics, it doesn't look as robust feature.

misteraddons commented:
What part of the cd image is already CRC32'd, and does it work across bin/cue and chd formats? The purpose is to advertise the current game along with the current core for scripts and third party utilities to display metadata.

We agree that cartridges are easy. Maybe we start with file path and cartridge CRC32?

I'm mostly following Robert's idea with the PSX and N64 core, both of which extract product code. I know the N64 core uses the local database to determine save time and region. In the future, could main set the region for cores which don't have auto region?

sorgelig commented:
> Maybe we start with file path and cartridge CRC32?

it's a;ready done. That's how cheat engine works.

misteraddons commented:
What’s the minimal code needed to reliably identify CD games?

Current CD Core Behavior:
- PSX: Uses product code (from disc) or filename for cheats.
- TurboGrafx-CD & Mega CD: Cheats use filename only; no product code or CRC32.
- Saturn, CD-i (WIP), Neo Geo CD: No cheats or product code support yet.

Current Game Identification Use Cases:
- N64: Uses N64-database.txt to set region, save type, etc.
- PSX: Uses region to select the correct BIOS

Future Game Identification Use Cases:
- TurboGrafx-CD: Auto-enable Arcade Card.
- Saturn Auto RAM cart/region settings - I need to check if Saturn still has region specific boot roms.
- Saturn/CD-i: Game ID might help cheat support (CD-i has 2 cheats haha).

Using CRC32 on the first 100 sectors, combined with product code extraction is fast, deterministic, and works with both .bin/.cue and .chd formats. 

Example ID timing (first 100 sectors):
```
Saturn Product Code: T14310G (0 ms)
Saturn CRC32: 300C0682 (4 ms)
```

Some platforms (e.g., TurboGrafx-CD) lack product codes entirely, and others may share or miss them—making CRC32 useful for differentiation. While CRC32 could be a universal fallback, product codes remain the standard.

For /tmp/GAMENAME, we could start with filename + CRC32 or product code where available.

Want to see the updated code with a common CRC32 sector function and per-core product code extraction code?

misteraddons committed:
Remove CD product code lookup, and crc32

misteraddons committed:
Remove product codes

/tmp/GAMENAME only uses file path and crc32

misteraddons committed:
Clean up

misteraddons committed:
Cleanup

misteraddons committed:
Only write CRC32 and or Product Code to /tmp/GAMEID

Realizing we already have the file path, this change writes the already calculated CRC32 and or game ID (N64 and PSX only) to /tmp/GAMEID. This ignores boot*.rom and cd_bios*.rom

CRC32: XXXXXXXX
Product Code: XXXXXXXXXXXX

misteraddons committed:
Clean up

misteraddons committed:
Update neogeo_loader.cpp

misteraddons committed:
Skip GAMEID for all BIOS files

misteraddons committed:
Guard behind log_file_entry ini option and clean up print statements

misteraddons committed:
Add MRA GAMEID support and improve GAMEID writing

MRA files use setname as serial

misteraddons committed:
Add MGL handling for /tmp/ STARTPATH, FULLPATH, CURRENTPATH, and FILESELECT

misteraddons committed:
Update mgl /tmp/ file handling

misteraddons commented:
Any feedback?
